src/nec-usbatermfw.c

```diff
@@ -0,0 +1,328 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+/*
+ * Copyright (C) 2024 INAGAKI Hiroshi <musashino.open@gmail.com>
+ */
+
+#include <byteswap.h>
+#include <endian.h>
+#include <errno.h>
+#include <libgen.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+
+#define FWHDR_LEN		0x20
+#define BLKHDR_LEN		0x18
+
+#define BLKHDR_FLAGS_GZIP	0x80000000
+#define BLKHDR_FLAGS_EXEC	0x00000002
+#define BLKHDR_FLAGS_SYS	0x00000001
+#define BLKHDR_DEF_LOADADDR	0x80040000
+#define BLKHDR_DEF_ENTRYP	BLKHDR_DEF_LOADADDR
+
+#define FILETYPE_LEN		0x20
+
+#define PADBLK_LEN		0x4
+#define DATABLK_CNT_MAX		2
+
+#if (__BYTE_ORDER == __LITTLE_ENDIAN)
+#  define HOST_TO_LE16(x)	(x)
+#  define HOST_TO_LE32(x)	(x)
+#  define HOST_TO_BE16(x)	bswap_16(x)
+#  define HOST_TO_BE32(x)	bswap_32(x)
+#else
+#  define HOST_TO_LE16(x)	bswap_16(x)
+#  define HOST_TO_LE32(x)	bswap_32(x)
+#  define HOST_TO_BE16(x)	(x)
+#  define HOST_TO_BE32(x)	(x)
+#endif
+
+struct header
+{
+	uint32_t flags;
+	uint32_t totlen;
+	uint32_t hdrlen;
+	uint32_t cksum;
+	uint32_t loadaddr;
+	uint32_t entryp;
+} __attribute__ ((packed));
+
+static char *progname;
+FILE *outbin;
+int nblk = 0;
+
+static void usage(void)
+{
+	printf("Usage: %s <output> [OPTIONS...]\n", progname);
+	printf("\n"
+	       "Options:\n"
+	       "  -t <type>       set firmware type to <type>\n"
+	       "  -f <flags>      set data flags to <flags>\n"
+	       "  -a <loadaddr>   set data load address to <loadaddr>\n"
+	       "  -e <entry>      set data entry point to <entry>\n"
+	       "  -d <file>       read input data from file <file> (max. %dx)\n",
+	       DATABLK_CNT_MAX);
+}
+
+static int strtou32(char *arg, unsigned int *val)
+{
+	char *endptr = NULL;
+
+	errno = 0;
+	*val = strtoul(arg, &endptr, 0);
+	return (errno || (endptr && *endptr)) ? -EINVAL : 0;
+}
+
+static int write_data(const char *buf, size_t length, uint32_t *cksum)
+{
+	size_t padlen = 0;
+	size_t pos;
+
+	if (length % PADBLK_LEN)
+		padlen = PADBLK_LEN - length % PADBLK_LEN;
+
+	if (fwrite(buf, 1, length, outbin) != length) {
+		fprintf(stderr, "Couldn't write %zu bytes\n", length);
+		return -EACCES;
+	}
+
+	if (cksum)
+		for (pos = 0; pos < length; pos += 2) {
+			if (length - pos == 1)
+				*cksum += HOST_TO_LE16(buf[pos]);
+			else
+				*cksum += HOST_TO_LE16(*(uint16_t *)(buf + pos));
+		}
+
+	if (padlen) {
+		int i;
+		for (i = 0; i < padlen; i++)
+			fputc(0x0, outbin);
+	}
+
+	return padlen;
+}
+
+static int write_blockheader(uint32_t flags,
+			     uint32_t loadaddr, uint32_t entryp,
+			     size_t datalen, int padlen, uint32_t cksum)
+{
+	struct header hdr;
+	void *cur = &hdr;
```

Comment:
Why not make this a `uint16_t` pointer?
```suggestion
	uint16_t *cur = (uint16_t *)&hdr;
```
Then you need less casts in the for loop.